### PR TITLE
removeMetadata

### PR DESCRIPTION
**Pull request by bourgeoa**
`store.remove()` fails on triple containing a `Collection`

### CODE REVIEW

File: src/store.ts [hunk 1]
```diff
@@ -871,25 +871,46 @@ export default class IndexedFormula extends Formula { // IN future - allow pass
...
-  removeDocument(doc: Quad_Graph): IndexedFormula {
-    const meta = this.sym('chrome://TheCurrentSession') // or this.rdfFactory.namedNode('chrome://TheCurrentSession')
-    const linkNamespaceURI = 'http://www.w3.org/2007/ont/link#' // alain
-    // remove request/response and metadata
-    const requests = this.statementsMatching(undefined, this.sym(`${linkNamespaceURI}requestedURI`), this.rdfFactory.literal(doc.value), meta).map(st => st.subject)
+  removeMetadata(doc: Quad_Graph): IndexedFormula {
+    const meta = this.fetcher?.appNode // this.sym('chrome://TheCurrentSession')
+    const linkNamespaceURI = 'http://www.w3.org/2007/ont/link#'
+    const kb = this
+    // removeMatches() --> removeMany() --> remove() fails on Collection
```
Comment (angelo-v) [inline src/store.ts:881]: Wouldn't it be better to fix the remove() function to work with Collection instead of applying a workarround here?

File: src/store.ts [hunk 2]
```diff
@@ -871,25 +871,46 @@ export default class IndexedFormula extends Formula { // IN future - allow pass
...
+      for (var i = 0; i < sts.length; i++) {
+        kb.removeStatement(sts[i])
+      }
+    }
+    const requests = this.statementsMatching(null, this.sym(`${linkNamespaceURI}requestedURI`), this.rdfFactory.literal(doc.value), meta).map(st => st.subject)
     for (var r = 0; r < requests.length; r++) {
       const request = requests[r]
-      if (request !== undefined) {
-        this.removeMatches(request, null, null, meta)
+      if (request != null) { // null or undefined
```
Comment (angelo-v) [inline src/store.ts:892]: The comment says "or undefined" but the code onyl checks for null. Why?

File: tests/unit/update-manager-test.js [hunk 3]
```diff
@@ -223,6 +223,17 @@ describe('UpdateManager', () => {
       expect(updater.editable(doc1)).to.equal(undefined)
     })
 
+    it('Should not detect a document is editable from metadata after removeMetadata', () => {
```
Comment (angelo-v) [inline tests/unit/update-manager-test.js:226]: How are these tests related to the changes? I would have expected a test that assures that Collections are now working.

File: src/store.ts (hunk 2 above)
Comment (bourgeoa) [reply src/store.ts:892]: Loose inequality does it
It is used in other places in rdflib

File: tests/unit/update-manager-test.js (hunk 3 above)
Comment (bourgeoa) [reply tests/unit/update-manager-test.js:226]: They are not. It only test the new function `removeMetadata`

File: src/store.ts (hunk 1 above)
Comment (bourgeoa) [reply src/store.ts:881]: Yes it would. I did not find howto.

Comment (timbl) [reply src/store.ts:881]: What goes wrong with when one f the nodes being removed is a collection?

Comment (bourgeoa) [reply src/store.ts:881]: @timbl
This is I hope a detailed enough issue https://github.com/linkeddata/rdflib.js/issues/631

File: tests/unit/update-manager-test.js (hunk 3 above)
Comment (bourgeoa) [reply tests/unit/update-manager-test.js:226]: the remove date includes the tricky Collection.
see issue https://github.com/linkeddata/rdflib.js/issues/631

File: src/store.ts (hunk 2 above)
Comment (jeff-zucker) [reply src/store.ts:893]: **** I am not sure I understand this - if I create a file containing only `if(foo != null) console.log(1); ` I get `ReferenceError: foo is not **defined**`.  So I don't see how this can work.  I so much miss perl in this regard where you could just say if(foo) and that would cover null, undefined, and empty.

**Comment by jeff-zucker** [comment]
AH! Thank you.  It works if the variable is declared but not defined but
does not work if the variable is not declared, makes sense.

**Comment by jeff-zucker** [comment]
I ran the attached script on solidcommunity.net and it errored with `statement  to be removed is not in store`.  I got this error even though there was no collection so I don't think the collection is the problem.

The same script on the test server did not produce an error.  But it also did not remove all metadata about the document.  Even after the new removeDocument(), there were three statements left over.  The predicates related to the document that were not removed were acl, describedBy, and type - things from the link relations.  

I am not sure if it was intentional to leave these, but if not, it should be easy to fix.

```turtle

<script src="https://solidcommunity.net:8443/mashlib.js"></script>                                  
<script type="module">                                                                              
                                                                                                    
let doc = UI.rdf.sym("https://localhost:8443/public/s/test/tmpr/test.ttl");                         
                                                                                                    
async function test(){                                                                              
  await UI.store.fetcher.load(doc);                                                                 
  show('load,store');                                                                               
  await UI.store.removeDocument(doc);                                                                     
  show('load,store,removeDocument');;                                                               
}                                                                                                   
test();                                                                                             
                                                                                                    
function show(msg){                                                                                 
  let stmts  = UI.store.match(null,null,null,doc);                                                  
  console.log(`${stmts.length} statements with graph specified after ` + msg);                      
  stmts  = UI.store.match();                                                                        
  console.log(`${stmts.length} statements with graph not specified after ` + msg);                  
}                                                                                                   
                                                                                                    
</script>                                                                                           
```
The document in question looks like this:
```text
<#A> <#isa> <#B>.                                                                                   
<#C> <#isa> <#D>.                                                                                   
```
The script output was this (notice the 3 in the last one) : 
```text
2 statements with graph specified after load,store
30 statements with graph not specified after load,store
0 statements with graph specified after load,store,removeDocument
3 statements with graph not specified after load,store,removeDocument
```
```
```

**Comment by jeff-zucker** [comment]
Also note that removeDocument did remove everything that had the document as the graph, but the three which were not  removed had the document as subject and a blank node as the graph.

**Comment by bourgeoa** [comment]
> I ran the attached script on solidcommunity.net and it errored with `statement to be removed is not in store`. I got this error even though there was no collection so I don't think the collection is the problem.

The collection is in the metadata graph that is were the collection is an issue, not on the document graph

> Also note that `removeDocument` did remove everything that had the document as the graph, but the three which were not removed had the document as subject and a blank node as the graph.

Then these 3 statements are not related to either the graph document or the document part of the metadata graph
I shall look where these are created. If they may be removed. They make no harm for the time being

**Review by jeff-zucker — APPROVED** [review]
I'll raise a separate issue on the un-removed link relations.  Other than that, this looks good to go.
